ddl/backfilling.go

```diff
@@ -520,107 +520,144 @@ func splitTableRanges(t table.PhysicalTable, store kv.Storage, startKey, endKey
 	return ranges, nil
 }
 
-func waitTaskResults(scheduler *backfillScheduler, batchTasks []*reorgBackfillTask,
-	totalAddedCount *int64) (kv.Key, int64, error) {
-	var (
-		firstErr   error
-		addedCount int64
-	)
-	keeper := newDoneTaskKeeper(batchTasks[0].startKey)
-	taskSize := len(batchTasks)
-	for i := 0; i < taskSize; i++ {
-		result := <-scheduler.resultCh
-		if result.err != nil {
-			if firstErr == nil {
-				firstErr = result.err
-			}
-			logutil.BgLogger().Warn("[ddl] backfill worker failed",
-				zap.String("result next key", hex.EncodeToString(result.nextKey)),
-				zap.Error(result.err))
-			// Drain tasks.
-			cnt := drainTasks(scheduler.taskCh)
-			// We need to wait all the tasks to finish before closing it
-			// to prevent send on closed channel error.
-			taskSize -= cnt
-			continue
-		}
-		*totalAddedCount += int64(result.addedCount)
-		addedCount += int64(result.addedCount)
-		keeper.updateNextKey(result.taskID, result.nextKey)
-		if i%scheduler.workerSize()*4 == 0 {
-			// We try to adjust the worker size regularly to reduce
-			// the overhead of loading the DDL related global variables.
-			err := scheduler.adjustWorkerSize()
-			if err != nil {
-				logutil.BgLogger().Warn("[ddl] cannot adjust backfill worker size", zap.Error(err))
-			}
-		}
-	}
-	return keeper.nextKey, addedCount, errors.Trace(firstErr)
+type resultConsumer struct {
+	dc      *ddlCtx
+	wg      *sync.WaitGroup
+	taskCnt *atomic.Int64
+	doneCh  chan struct{}
+	err     *atomic.Value
```

Comment:
atomic.Value is replaced by atomic.Pointer to avoid type convert.

Reply:
Removed.